src/query/functions/src/scalars/conditionals/in_basic.rs

```diff
@@ -79,16 +81,66 @@ macro_rules! scalar_contains {
     }};
 }
 
+macro_rules! bool_contains {
+    ($T: ident, $INPUT_COL: expr, $ROWS: expr, $COLUMNS: expr, $CAST_TYPE: ident, $FUNC_CTX: expr) => {{
+        let mut builder: ColumnBuilder<bool> = ColumnBuilder::with_capacity($ROWS);
+        let mut vals = 0;
+        for col in &$COLUMNS[1..] {
+            let col = cast_column_field(col, col.data_type(), &$CAST_TYPE, &$FUNC_CTX)?;
+            let col_viewer = $T::try_create_viewer(&col)?;
+            if col_viewer.valid_at(0) {
+                let val = col_viewer.value_at(0);
+                vals |= 1 << (val as u8 + 1);
+            }
+        }
+        let input_viewer = $T::try_create_viewer(&$INPUT_COL)?;
+        for (row, val) in input_viewer.iter().enumerate() {
+            let contains = ((vals >> (val as u8 + 1)) & 1) > 0;
+            let valid = input_viewer.valid_at(row);
+            builder.append(valid && ((contains && !NEGATED) || (!contains && NEGATED)));
+        }
+        return Ok(builder.build($ROWS));
+    }};
+}
+
+macro_rules! string_contains {
+    ($T: ident, $INPUT_COL: expr, $ROWS: expr, $COLUMNS: expr, $CAST_TYPE: ident, $FUNC_CTX: expr) => {{
+        let mut builder: ColumnBuilder<bool> = ColumnBuilder::with_capacity($ROWS);
+        let mut vals_set: HashSetWithStackMemory<64, KeysRef> = HashSetWithStackMemory::create();
+        let mut inserted = false;
+        for col in &$COLUMNS[1..] {
+            let col = cast_column_field(col, col.data_type(), &$CAST_TYPE, &$FUNC_CTX)?;
+            let col_viewer = $T::try_create_viewer(&col)?;
+            if col_viewer.valid_at(0) {
+                let val = col_viewer.value_at(0);
+                let key = KeysRef::create(val.as_ptr() as usize, val.len());
+                vals_set.insert_key(&key, &mut inserted);
+            }
+        }
+        let input_viewer = $T::try_create_viewer(&$INPUT_COL)?;
+        for (row, val) in input_viewer.iter().enumerate() {
```

Comment:
if `INPUT_COL ` is a const-column,  we could erase the iter() ?

Reply:
this may happens in hive storage,  the partition column is virtual and const

Reply:
`input_viewer` already handles the constant cases.

See:

```
 #[inline]
    fn value_at(&self, index: usize) -> bool {
        self.values.get_bit(index & self.non_const_mask)
    }
```

Reply:
cool !

Reply:
correct me if I mis-understanding,  if INPUT_COL is const-column, we could rewrite code like this, seems more effective.
```
    let val = input_viewer.value_at(0);
    let contains = vals_set.contains(val);
    let valid = input_viewer.valid_at(0);
  // construct a bool const collumn
    return constColumn::new(contains,  xx)
```

Reply:
Yes, so we set `self.non_const_mask` to zero for const column, the index will always be zero after `index & 0`.